### PR TITLE
build(deps): bump grpc version to 1.58.0 and protobuf to 3.24.0

### PR DESCRIPTION
**Pull request by mpeddada1**
Notable changes in this PR:
- Upgrade of protobuf to 3.24.0.
- Upgrade of grpc to 1.58.0.
- Upgrade of errorprone to 2.20.0.
- Re-generation of showcase golden files with new versions of protobuf and grpc.
- Pinned GA job for BOM validation to commit sha of https://github.com/googleapis/java-cloud-bom/pull/6191 (which ignores grpc-binder, grpc-cronet and grpc-android).
- Modified the GA job for downstream testing to use HEAD of google-cloud-java to account for changes in https://github.com/googleapis/google-cloud-java/pull/9854

cc/ @suztomo

### CODE REVIEW

**Comment by mpeddada1** [comment]
Stacktrace:
```
Error:  +-com.google.api.grpc:grpc-google-common-protos:2.24.1-SNAPSHOT
Error:    +-com.google.protobuf:protobuf-java:3.23.2
Error:  and
Error:  +-com.google.api.grpc:grpc-google-common-protos:2.24.1-SNAPSHOT
Error:    +-io.grpc:grpc-protobuf:1.58.0
Error:      +-com.google.protobuf:protobuf-java:3.23.2 (managed) <-- com.google.protobuf:protobuf-java:3.24.0
```
gRPC 1.58.0 needs protobuf 3.24.0

**Comment by suztomo** [comment]
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.3.0:enforce (enforce) on project gax-grpc: 
Error:  Rule 2: org.apache.maven.enforcer.rules.dependency.RequireUpperBoundDeps failed with message:
Error:  Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Error:  Require upper bound dependencies error for com.google.errorprone:error_prone_annotations:2.18.0 paths to dependency are:
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-com.google.api:api-common:2.16.1-SNAPSHOT
Error:      +-com.google.errorprone:error_prone_annotations:2.18.0
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-api:1.58.0
Error:      +-com.google.errorprone:error_prone_annotations:2.20.0
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-stub:1.58.0
Error:      +-com.google.errorprone:error_prone_annotations:2.20.0 [runtime]
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-netty-shaded:1.58.0 [runtime]
Error:      +-com.google.errorprone:error_prone_annotations:2.20.0 [runtime]
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-com.google.auth:google-auth-library-oauth2-http:1.19.0
Error:      +-com.google.http-client:google-http-client:1.43.3 (managed) <-- com.google.http-client:google-http-client:1.42.3
Error:        +-com.google.errorprone:error_prone_annotations:2.18.0
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-alts:1.58.0
Error:      +-io.grpc:grpc-grpclb:1.58.0 (managed) <-- io.grpc:grpc-grpclb:1.58.0
Error:        +-com.google.errorprone:error_prone_annotations:2.20.0 [runtime]
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-alts:1.58.0
Error:      +-io.grpc:grpc-grpclb:1.58.0 (managed) <-- io.grpc:grpc-grpclb:1.58.0
Error:        +-com.google.protobuf:protobuf-java-util:3.24.0 [runtime] (managed) <-- com.google.protobuf:protobuf-java-util:3.24.0 [runtime]
Error:          +-com.google.errorprone:error_prone_annotations:2.18.0 [runtime]
Error:  and
Error:  +-com.google.api:gax-grpc:2.33.1-SNAPSHOT
Error:    +-io.grpc:grpc-googleapis:1.58.0 [runtime]
Error:      +-io.grpc:grpc-xds:1.58.0 [runtime] (managed) <-- io.grpc:grpc-xds:1.58.0 [runtime]
Error:        +-io.grpc:grpc-services:1.58.0 [runtime] (managed) <-- io.grpc:grpc-services:1.58.0 [runtime]
Error:          +-com.google.errorprone:error_prone_annotations:2.20.0 [runtime]
Error:  ]
```

**Comment by mpeddada1** [comment]
Thanks Tomo! Making some progress. Running into a different error now:
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project gax-grpc: Compilation failure: Compilation failure: 
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/testing/InProcessServer.java:[40,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/GrpcDirectServerStreamingCallableTest.java:[54,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/GrpcDirectStreamingCallableTest.java:[52,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/testing/MockServiceHelper.java:[36,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/testing/LocalChannelProvider.java:[48,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/GrpcCallableFactoryTest.java:[53,25] package io.grpc.inprocess does not exist
Error:  /home/runner/work/sdk-platform-java/sdk-platform-java/gax-java/gax-grpc/src/test/java/com/google/api/gax/grpc/GrpcResponseMetadataTest.java:[50,25] package io.grpc.inprocess does not exist
```

**Comment by mpeddada1** [comment]
From the GRPC Github Releases page; grpc-inprocess and grpc-util have been moved out of grpc-core to be their own artifacts.
```
core, inprocess, util: 2 new artifacts grpc-inprocess and grpc-util have been created by moving code from grpc-core to facilitate Java module support (https://github.com/grpc/grpc-java/pull/10362, https://github.com/grpc/grpc-java/pull/10390)
```

**Comment by mpeddada1** [comment]
Showcase golden files were updated but we are still running into:
```
diff -ru /home/runner/work/sdk-platform-java/sdk-platform-java/showcase/scripts/../grpc-gapic-showcase-v1beta1/src/main/java/com/google/showcase/v1beta1/ComplianceGrpc.java /home/runner/work/sdk-platform-java/sdk-platform-java/showcase/scripts/output/showcase-output/grpc-showcase-output/src/main/java/com/google/showcase/v1beta1/ComplianceGrpc.java
--- /home/runner/work/sdk-platform-java/sdk-platform-java/showcase/scripts/../grpc-gapic-showcase-v1beta1/src/main/java/com/google/showcase/v1beta1/ComplianceGrpc.java	2023-09-15 22:43:21.740534658 +0000
+++ /home/runner/work/sdk-platform-java/sdk-platform-java/showcase/scripts/output/showcase-output/grpc-showcase-output/src/main/java/com/google/showcase/v1beta1/ComplianceGrpc.java	2023-09-15 22:45:08.711177045 +0000
@@ -17,7 +17,7 @@
 
   private ComplianceGrpc() {}
 
-  public static final String SERVICE_NAME = "google.showcase.v1beta1.Compliance";
+  public static final java.lang.String SERVICE_NAME = "google.showcase.v1beta1.Compliance";
 
   // Static method descriptors that strictly reflect the proto.
   private static volatile io.grpc.MethodDescriptor<com.google.showcase.v1beta1.RepeatRequest,
@@ -1124,9 +1124,9 @@
   private static final class ComplianceMethodDescriptorSupplier
       extends ComplianceBaseDescriptorSupplier
       implements io.grpc.protobuf.ProtoMethodDescriptorSupplier {
-    private final String methodName;
+    private final java.lang.String methodName;
 
-    ComplianceMethodDescriptorSupplier(String methodName) {
+    ComplianceMethodDescriptorSupplier(java.lang.String methodName) {
       this.methodName = methodName;
     }
```

**Comment by mpeddada1** [comment]
`ci/gapic-generator-java-bom` failing with:
```
Error:  Failed to execute goal on project bom-validation-canary-project: Could not resolve dependencies for project com.google.cloud:bom-validation-canary-project:jar:0.0.1-SNAPSHOT: The following artifacts could not be resolved: io.grpc:grpc-android:jar:1.58.0, io.grpc:grpc-binder:jar:1.58.0, androidx.annotation:annotation:jar:1.6.0, androidx.core:core:jar:1.10.1, androidx.lifecycle:lifecycle-common:jar:2.6.1, io.grpc:grpc-cronet:jar:1.58.0, org.chromium.net:cronet-api:jar:108.5359.79: Could not find artifact io.grpc:grpc-android:jar:1.58.0 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
Error:  
```

**Comment by suztomo** [comment]
Can you remove unnecessary artifacts from the canary project?

**Comment by mpeddada1** [comment]
> Can you remove unnecessary artifacts from the canary project?

Thanks for the suggestion, Tomo. Done in https://github.com/googleapis/java-cloud-bom/pull/6189

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -282,7 +282,7 @@ jobs:
       run: |
         mvn install -B -ntp -DskipTests -Dclirr.skip -Dcheckstyle.skip
     - name: Validate gapic-generator-java-bom
-      uses: googleapis/java-cloud-bom/tests/validate-bom@v26.13.0
+      uses: googleapis/java-cloud-bom/tests/validate-bom
```
Comment (suztomo) [inline .github/workflows/ci.yaml:285]: In the document https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iduses I don't see the "ref" after "@" is optional.

**Comment by mpeddada1** [comment]
Removal of [MetadataUtils#attachHeader](https://github.com/grpc/grpc-java/blob/cde208afce38186d06084605abed36f21b003465/stub/src/main/java/io/grpc/stub/MetadataUtils.java#L49-L57) is causing compilation errors in[ ITKmsTest.java](https://github.com/googleapis/google-cloud-java/blob/c000864ab4566b1e260119fd959d35e0641fc6f7/java-kms/google-cloud-kms/src/test/java/com/google/cloud/kms/it/ITKmsTest.java#L88).

File: .kokoro/presubmit/downstream-compatibility.sh [hunk 2]
```diff
@@ -38,9 +38,14 @@ echo "Install complete. java-shared-dependencies = $SHARED_DEPS_VERSION"
...
-  last_release=$(find_last_release_version "$repo")
-  git clone "https://github.com/googleapis/$repo.git" --depth=1 --branch "v$last_release"
+  # Perform testing on HEAD for google-cloud-java
+  if [ $repo == "google-cloud-java" ]; then
+    git clone "https://github.com/googleapis/$repo.git" --depth=1
+  else
+    # Perform testing on last release, not HEAD
+    last_release=$(find_last_release_version "$repo")
+    git clone "https://github.com/googleapis/$repo.git" --depth=1 --branch "v$last_release"
+  fi
```
Comment (mpeddada1) [inline .kokoro/presubmit/downstream-compatibility.sh:48]: The downstream checks for java-kms were failing with a compilation error due to a removal of a method in grpc-stub:1.58.0.  Needed to make this update to verify the changes in the monorepo: https://github.com/googleapis/google-cloud-java/pull/9854.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -282,7 +282,7 @@ jobs:
       run: |
         mvn install -B -ntp -DskipTests -Dclirr.skip -Dcheckstyle.skip
     - name: Validate gapic-generator-java-bom
-      uses: googleapis/java-cloud-bom/tests/validate-bom@v26.13.0
+      uses: googleapis/java-cloud-bom/tests/validate-bom@2414e159ab8f882717ab7f6ae4c4a02c096b94fa
```
Comment (mpeddada1) [inline .github/workflows/ci.yaml:285]: To test with changes included in https://github.com/googleapis/java-cloud-bom/pull/6191

**Comment by mpeddada1** [comment]
We've not counted golden file coverage as a blocker for merging this PR in the past (ex, https://github.com/googleapis/sdk-platform-java/pull/1755). However, in the future, it may be helpful to re-evaluate the use of our Sonarcloud coverage. 
cc/ @burkedavison

**Review by suztomo — APPROVED** [review]

**Pull request merged**